version/cmp.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package version
+
+import (
+	"log"
+	"strconv"
+	"strings"
+)
+
+// AtLeast returns whether version is at least the specified minimum
+// version.
+//
+// Version comparison in Tailscale is a little complex, because we
+// switched "styles" a few times, and additionally have a completely
+// separate track of version numbers for OSS-only builds.
+//
+// AtLeast acts conservatively, returning true only if it's certain
+// that version is at least minimum. As a result, it can produce false
+// negatives, for example when an OSS build supports a given feature,
+// but AtLeast is called with an official release number as the
+// minimum
+//
+// version and minimum can both be either an official Tailscale
+// version numbers (major.minor.patch-extracommits-extrastring), or an
+// OSS build datestamp (date.YYYYMMDD). For Tailscale version numbers,
+// AtLeast also accepts a prefix of a full version, in which case all
+// missing fields are assumed to be zero.
+func AtLeast(version string, minimum string) bool {
```

Comment:
Being in this package, I assume this compared against the current version.

Maybe add `func Cmp(a, b string) int` and then make AtLeast only take a String?

Reply:
Also fine as-is. Maybe renamed? Reads a bit odd as-is.

Reply:
So, I started out with a Cmp, but the problem is that our version numbers don't have a single total order. There is no return value for Cmp("date.1234", "0.98") that makes sense.

Instead, I tried to make a function that is scoped to specifically "can you tell me for sure that version is > minimum? It's okay to say no if unsure." Such a function can't be used to sort an array, only to do stuff like version gating (which is what I need this for).

Re: package: I need this in the control server to compare client-reported versions, so I can't make the comparison implicit against the burned-in version. But it's still utilities for manipulating version numbers, so this feels like it should be the place for it?

Dunno. I can also make it into a corp-only helper package, if you prefer that.

Reply:
Going to submit as-is so I can work on the control server, but happy to figure out a better place/name for this. I agree that a binary comparator is a bit weird within this package.